### PR TITLE
Use bytes in parquet rather than custom Buffer implementation (#1474)

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Part of #1474.

# Rationale for this change
 
See ticket, in particular I want to use this as part of #1605 

# What changes are included in this PR?

This replaces `Buffer` with `Vec`, it also make `ByteBufferPtr` use `Bytes` internally. A follow up PR will replace `ByteBufferPtr` with `Bytes` but I wanted to avoid making this PR too large

# Are there any user-facing changes?

Technically this only makes changes to experimental APIs, but it probably constitutes a breaking change

### CODE REVIEW

File: parquet/src/util/memory.rs [hunk 1]
```diff
@@ -17,532 +17,119 @@
...
-
-    /// Updates this buffer with new `start` position and length `len`.
-    ///
-    /// Range should be within current start position and length.
-    #[inline]
-    pub fn set_range(&mut self, start: usize, len: usize) {
-        assert!(self.start <= start && start + len <= self.start + self.len);
-        self.start = start;
-        self.len = len;
-    }
```
Comment (tustvold) [inline parquet/src/util/memory.rs:317]: These weren't actually being used anywhere, and they have somewhat strange semantics as they don't take into account the pre-existing offset.

File: parquet/src/column/reader.rs [hunk 2]
```diff
@@ -460,20 +460,20 @@ fn parse_v1_level(
     num_buffered_values: u32,
     encoding: Encoding,
     buf: ByteBufferPtr,
-) -> Result<ByteBufferPtr> {
+) -> Result<(usize, ByteBufferPtr)> {
```
Comment (alamb) [inline parquet/src/column/reader.rs:463]: I recommend documenting in comments what this `usize` is -- namely the number of bytes that was read?

File: parquet/src/util/memory.rs [hunk 3]
```diff
@@ -17,532 +17,119 @@
...
-    mem,
-    ops::{Index, IndexMut},
-    sync::{
-        atomic::{AtomicI64, Ordering},
-        Arc, Weak,
-    },
+    ops::Index,
 };
 
-// ----------------------------------------------------------------------
```
Comment (alamb) [inline parquet/src/util/memory.rs:31]: I think highlighting the loss of `MemTracker` in this change is probably important for anyone who is using it currently

**Review by alamb — APPROVED** [review]
Looks good to me from a code and API level. I like this change

I am a little concerned about the loss of `MemoryTracker` API (as I have no idea who is using that API)

DataFusion does not seem to https://github.com/apache/arrow-datafusion/search?q=MemoryTracker

cc @nevi-me @sunchao @jhorstmann @bjchambers - it would also be great to ping anyone else who people may remember being interested in this area of parquet 

It seems to have a non trivial number of dependents these days: https://crates.io/crates/parquet/reverse_dependencies

**Review by sunchao — COMMENTED** [review]
The `MemTracker ` is a half-baked thing and I'm fine with removing it. Eventually, though, I think it'd still be nice to have a way to track memory usage in the parquet path. Perhaps we can have some memory manager that can be passed in from outside when reading or writing Parquet, which is responsible for allocating & recycling byte buffers, but I'm not sure how it can integrate nicely with the `Bytes` changes.

**Comment by tustvold** [comment]
My reasoning for not worrying about removing Memtracker is I made the API experimental a while back, and there haven't been any complaints.

I can keep it if people feel strongly, I'm just trying to reduce the amount of custom code we have 😅

**Comment by alamb** [comment]
> My reasoning for not worrying about removing Memtracker is I made the API experimental a while back, and there haven't been any complaints.

I am all for deleting it, especially with some evidence that it isn't used 👍

**Review by sunchao — APPROVED** [review]
+1. I think we can make more changes based on this in future if we need to add a memory manager.

**Pull request merged**